go/pkg/eventstream/consumer.go

```diff
@@ -0,0 +1,223 @@
+// event. The method returns immediately after starting the background processing.
+//
+// Single-handler enforcement:
+//
+//	This method can only be called once per consumer instance. Subsequent calls
+//	are silently ignored to prevent multiple competing handlers and race conditions.
+//	This design ensures clear ownership of message processing.
+//
+// Handler function:
+//
+//	The handler is called for each received event with a context that has a 30-second
+//	timeout. If the handler returns an error, the error is logged but message
+//	processing continues. Handler errors do not cause the consumer to stop.
+//
+// Message processing guarantees:
+//   - At-least-once delivery (messages may be redelivered on failure)
+//   - Messages from the same partition are processed in order
+//   - Automatic offset commits for successfully processed messages
+//   - Consumer group rebalancing handles instance failures automatically
+//
+// Error handling:
+//
+//	All errors are logged rather than returned since this method runs asynchronously:
+//	- Kafka connection errors are logged and trigger automatic reconnection
+//	- Protobuf deserialization errors are logged and the message is skipped
+//	- Handler errors are logged but processing continues
+//	- Fatal errors (authentication, configuration) cause consumption to stop
+//
+// Performance characteristics:
+//   - Automatic message batching for improved throughput
+//   - Configurable consumer group for load balancing
+//   - Efficient protobuf deserialization with minimal allocations
+//   - Consumer group: "{topic}::{instanceID}" for instance-based load balancing
+//
+// Context handling:
+//
+//	The provided context is used for the entire consumption lifecycle. When the
+//	context is cancelled, the background goroutine stops and the consumer shuts down
+//	gracefully. Context cancellation is the primary mechanism for stopping consumption.
+//
+// Resource management:
+//
+//	The background goroutine automatically manages Kafka connections and consumer
+//	group membership. Call Close() when the consumer is no longer needed to ensure
+//	proper cleanup and consumer group departure.
+//
+// Example:
+//
+//	consumer := topic.NewConsumer()
+//
+//	// Start consuming in background
+//	consumer.Consume(ctx, func(ctx context.Context, event *MyEvent) error {
+//	    log.Printf("Received event: %+v", event)
+//	    // Process the event...
+//	    return nil // nil = success, error = logged but processing continues
+//	})
+//
+//	// Do other work while consuming happens in background...
+//
+//	// Clean shutdown
+//	consumer.Close()
+func (c *consumer[T]) Consume(ctx context.Context, handler func(context.Context, T) error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.subscribed {
+		// Already consuming, ignore subsequent calls
+		return
+	}
+
+	c.handler = handler
+	c.subscribed = true
+
+	readerConfig := kafka.ReaderConfig{
```

Comment:
we have to set StartOffset = kafka.LastOffset by default since we don't want new nodes to read all invalidations